feat(api-gateway): add meta about query in gql

nurulhudaapon opened the pull request:
**Check List**
- [x] Tests has been run in packages where changes made if available
- [x] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [x] Docs have been added / updated if required


**Description of Changes Made (if issue reference is not provided)**

This PR adds `extensions.meta` which is equivalent to all metadata about query result that is available via the REST API /load endpoint. The load endpoint returns metadata such as `total` and others but these were not available through the graphql API.

Here is an example of getting total/query-meta from GraphQL API: 
Query: 
```gql
query GetTotal {
  users: cube(total: true, limit: 1, meta: true) {
    users {
      email
    }
  }
  users_with_total: cube(total: true, limit: 1) {
    users {
      email
    }
  }
}
```
Response: 
```json
{
  "data": {
    "users": [
      {
        "users": {
          "email": "nurul@voyagesms.com"
        }
      }
    ],
    "users_with_total": [
      {
        "users": {
          "email": "nurul@voyagesms.com"
        }
      }
    ]
  },
  "extensions": {
    "meta": {
      "users": {
        "total": 5,
        "dbType": "postgres",
        "extDbType": "cubestore",
        "external": false,
        "lastRefreshTime": "2024-07-03T12:27:17.147Z",
        "slowQuery": false
      },
      "users_with_total": {
        "total": 5
      }
    }
  }
}
```

So this PR is enabling feature like getting the `total` via GraphQL API which wasn't possible previously and opens up adding other metadata if needed.

igorlukanin commented:
Hi @nurulhudaapon 👋 Thanks for your contribution!

I see that you not only add `resultMeta` but you also propagate `total`—that is amazing. Ideally, it would be great to have these changes in separate PRs.

Do I understand correctly that `resultMeta` would bring `meta` **together** with query results? That does seem like a deviation from what `/v1/meta` and `/v1/load` do for the REST API. Could you please elaborate on your design decisions here?

If the goal is to allow requesting `meta` via the GraphQL API, I would try to design it similarly to the REST API, i.e.g, have a dedicated way to request just `meta`, not pair it with a result set.

nurulhudaapon commented:
> Hi @nurulhudaapon 👋 Thanks for your contribution!
> 
> I see that you not only add `resultMeta` but you also propagate `total`—that is amazing. Ideally, it would be great to have these changes in separate PRs.
> 
> Do I understand correctly that `resultMeta` would bring `meta` **together** with query results? That does seem like a deviation from what `/v1/meta` and `/v1/load` do for the REST API. Could you please elaborate on your design decisions here?
> 
> If the goal is to allow requesting `meta` via the GraphQL API, I would try to design it similarly to the REST API, i.e.g, have a dedicated way to request just `meta`, not pair it with a result set.



> Hi @nurulhudaapon 👋 Thanks for your contribution!
> 
> I see that you not only add `resultMeta` but you also propagate `total`—that is amazing. Ideally, it would be great to have these changes in separate PRs.
> 
> Do I understand correctly that `resultMeta` would bring `meta` **together** with query results? That does seem like a deviation from what `/v1/meta` and `/v1/load` do for the REST API. Could you please elaborate on your design decisions here?
> 
> If the goal is to allow requesting `meta` via the GraphQL API, I would try to design it similarly to the REST API, i.e.g, have a dedicated way to request just `meta`, not pair it with a result set.

Hey 

> Hi @nurulhudaapon 👋 Thanks for your contribution!
> 
> I see that you not only add `resultMeta` but you also propagate `total`—that is amazing. Ideally, it would be great to have these changes in separate PRs.
> 
> Do I understand correctly that `resultMeta` would bring `meta` **together** with query results? That does seem like a deviation from what `/v1/meta` and `/v1/load` do for the REST API. Could you please elaborate on your design decisions here?
> 
> If the goal is to allow requesting `meta` via the GraphQL API, I would try to design it similarly to the REST API, i.e.g, have a dedicated way to request just `meta`, not pair it with a result set.

Hi @igorlukanin, sorry about the confusion about the /meta API, this is different and only adds metadata about query result. I've updated the PR title, description and also updated the doc. Let me know if it is clearer now. Thanks.

nurulhudaapon commented:
Hi @paveltiunov, sorry for tagging, just was looking if it's possible to get this in.

nurulhudaapon commented:
@igorlukanin This is kind of blocking us from adding pagination with single API call when using GraphQL, having this available sooner will be greatly appreciated.

nurulhudaapon commented:
@KSDaemon Just bumping to see if this is possible to get in. Thanks.
